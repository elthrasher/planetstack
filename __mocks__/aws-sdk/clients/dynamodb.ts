export const awsSdkPromiseResponse = jest.fn().mockReturnValue(Promise.resolve(true));

export const deleteFn = jest.fn().mockImplementation(() => ({ promise: awsSdkPromiseResponse }));

const getFn = jest.fn().mockImplementation(() => ({ promise: awsSdkPromiseResponse }));

const queryFn = jest.fn().mockImplementation(() => ({ promise: awsSdkPromiseResponse }));

export const putFn = jest.fn().mockImplementation(() => ({ promise: awsSdkPromiseResponse }));

export const updateFn = jest.fn().mockImplementation(() => ({ promise: awsSdkPromiseResponse }));

export class DocumentClient {
  delete = deleteFn;
  get = getFn;
  put = putFn;
  query = queryFn;
  update = updateFn;
  options = {};
}
