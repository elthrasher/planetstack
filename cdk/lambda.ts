import { Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Stack } from '@aws-cdk/core';

export const getFunctions = (scope: Stack): { [name: string]: LambdaFunction } => {
  const connectFn = new NodejsFunction(scope, 'ConnectFunction', {
    entry: `${__dirname}/../fns/connect.ts`,
    runtime: Runtime.NODEJS_14_X,
  });
  const disconnectFn = new NodejsFunction(scope, 'DisconnectFunction', {
    entry: `${__dirname}/../fns/disconnect.ts`,
    runtime: Runtime.NODEJS_14_X,
  });
  return { connectFn, disconnectFn };
};
