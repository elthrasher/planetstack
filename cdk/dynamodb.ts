import { AttributeType, BillingMode, Table } from '@aws-cdk/aws-dynamodb';
import { RemovalPolicy, Stack } from '@aws-cdk/core';

import { lambdaFunctions } from './lambda';

export const getTable = (stack: Stack, fns: lambdaFunctions): Table => {
  const table = new Table(stack, 'PlanetStackTable', {
    billingMode: BillingMode.PAY_PER_REQUEST,
    partitionKey: { name: 'pk', type: AttributeType.STRING },
    removalPolicy: RemovalPolicy.DESTROY,
    sortKey: { name: 'sk', type: AttributeType.STRING },
    tableName: `PlanetStackTable`,
    timeToLiveAttribute: 'exp',
  });
  table.grantReadWriteData(fns.addIcon);
  table.grantReadWriteData(fns.changeBackground);
  table.grantReadWriteData(fns.deleteIcon);
  table.grantReadData(fns.getState);
  table.grantReadWriteData(fns.moveIcon);
  table.grantWriteData(fns.onConnect);
  table.grantWriteData(fns.onDisconnect);

  return table;
};
