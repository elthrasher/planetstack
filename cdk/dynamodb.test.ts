import { SynthUtils } from '@aws-cdk/assert';
import { App, Stack } from '@aws-cdk/aws-sns/node_modules/@aws-cdk/core';

import { getTable } from './dynamodb';
import { getFunctions } from './lambda';

describe('DynamoDB', () => {
  test('returns a table', () => {
    const app = new App();
    const stack = new Stack(app, 'ApiTestStack', { env: { account: '123456789', region: 'us-east-1' } });
    getTable(stack, getFunctions(stack));
    const cfn = SynthUtils.toCloudFormation(stack);
    const resources = cfn.Resources;
    const matchObject: { Parameters: Record<string, unknown>; Resources: Record<string, unknown> } = {
      Parameters: expect.any(Object),
      Resources: {},
    };
    Object.keys(resources).forEach((res) => {
      switch (resources[res].Type) {
        case 'AWS::Lambda::Function':
          matchObject.Resources[res] = {
            Properties: { Code: expect.any(Object) },
          };
          break;
        default:
          break;
      }
    });

    expect(cfn).toMatchSnapshot(matchObject);
  });
});
