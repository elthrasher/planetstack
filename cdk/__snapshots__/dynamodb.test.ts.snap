// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamoDB returns a table 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "PlanetStackTable5EF25626": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "PlanetStackTable",
        "TimeToLiveSpecification": Object {
          "AttributeName": "exp",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "addIconFunction4EE4E627": Object {
      "DependsOn": Array [
        "addIconFunctionServiceRoleDefaultPolicy1634626F",
        "addIconFunctionServiceRole8C1DB12B",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "addIconFunctionServiceRole8C1DB12B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "addIconFunctionServiceRole8C1DB12B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "addIconFunctionServiceRoleDefaultPolicy1634626F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "addIconFunctionServiceRoleDefaultPolicy1634626F",
        "Roles": Array [
          Object {
            "Ref": "addIconFunctionServiceRole8C1DB12B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "changeBackgroundFunction2B84D2FD": Object {
      "DependsOn": Array [
        "changeBackgroundFunctionServiceRoleDefaultPolicy58777BF1",
        "changeBackgroundFunctionServiceRole679E1E93",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "changeBackgroundFunctionServiceRole679E1E93",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "changeBackgroundFunctionServiceRole679E1E93": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "changeBackgroundFunctionServiceRoleDefaultPolicy58777BF1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "changeBackgroundFunctionServiceRoleDefaultPolicy58777BF1",
        "Roles": Array [
          Object {
            "Ref": "changeBackgroundFunctionServiceRole679E1E93",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "deleteIconFunction316C8F7A": Object {
      "DependsOn": Array [
        "deleteIconFunctionServiceRoleDefaultPolicy22E6821B",
        "deleteIconFunctionServiceRole01CCD8F3",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "deleteIconFunctionServiceRole01CCD8F3",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "deleteIconFunctionServiceRole01CCD8F3": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deleteIconFunctionServiceRoleDefaultPolicy22E6821B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "deleteIconFunctionServiceRoleDefaultPolicy22E6821B",
        "Roles": Array [
          Object {
            "Ref": "deleteIconFunctionServiceRole01CCD8F3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "getStateFunction1C5566DE": Object {
      "DependsOn": Array [
        "getStateFunctionServiceRoleDefaultPolicyEC76E2CF",
        "getStateFunctionServiceRole7BBA9E9C",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "getStateFunctionServiceRole7BBA9E9C",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "getStateFunctionServiceRole7BBA9E9C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "getStateFunctionServiceRoleDefaultPolicyEC76E2CF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "getStateFunctionServiceRoleDefaultPolicyEC76E2CF",
        "Roles": Array [
          Object {
            "Ref": "getStateFunctionServiceRole7BBA9E9C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "moveIconFunctionB9241170": Object {
      "DependsOn": Array [
        "moveIconFunctionServiceRoleDefaultPolicy280FFE37",
        "moveIconFunctionServiceRole888B0F67",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "moveIconFunctionServiceRole888B0F67",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "moveIconFunctionServiceRole888B0F67": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "moveIconFunctionServiceRoleDefaultPolicy280FFE37": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "moveIconFunctionServiceRoleDefaultPolicy280FFE37",
        "Roles": Array [
          Object {
            "Ref": "moveIconFunctionServiceRole888B0F67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "onConnectFunction279CD46D": Object {
      "DependsOn": Array [
        "onConnectFunctionServiceRoleDefaultPolicy3FC4D257",
        "onConnectFunctionServiceRole23446B74",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "onConnectFunctionServiceRole23446B74",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "onConnectFunctionServiceRole23446B74": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "onConnectFunctionServiceRoleDefaultPolicy3FC4D257": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "onConnectFunctionServiceRoleDefaultPolicy3FC4D257",
        "Roles": Array [
          Object {
            "Ref": "onConnectFunctionServiceRole23446B74",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "onDisconnectFunctionD5542DF7": Object {
      "DependsOn": Array [
        "onDisconnectFunctionServiceRoleDefaultPolicy5F66930C",
        "onDisconnectFunctionServiceRoleAE265B92",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "onDisconnectFunctionServiceRoleAE265B92",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "onDisconnectFunctionServiceRoleAE265B92": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "onDisconnectFunctionServiceRoleDefaultPolicy5F66930C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PlanetStackTable5EF25626",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "onDisconnectFunctionServiceRoleDefaultPolicy5F66930C",
        "Roles": Array [
          Object {
            "Ref": "onDisconnectFunctionServiceRoleAE265B92",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
